File: java/src/test/java/bootcamp/FlowTests.java
package bootcamp;

import com.google.common.collect.ImmutableList;
import net.corda.core.concurrent.CordaFuture;
import net.corda.core.contracts.Command;
import net.corda.core.contracts.TransactionState;
import net.corda.core.transactions.SignedTransaction;
import net.corda.testing.node.MockNetwork;
import net.corda.testing.node.StartedMockNode;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import static net.corda.testing.node.NodeTestUtils.startFlow;
import static org.junit.Assert.assertEquals;

public class FlowTests {
    private MockNetwork network;
    private StartedMockNode a;
    private StartedMockNode b;

    @Before
    public void setup() {
        network = new MockNetwork(ImmutableList.of("bootcamp"));
        a = network.createPartyNode(null);
        b = network.createPartyNode(null);
        network.runNetwork();
    }

    @After
    public void tearDown() {
        network.stopNodes();
    }

    @Test
    public void transactionConstructedByFlowUsesTheCorrectNotary() throws Exception {
        TokenFlow flow = new TokenFlow(b.getInfo().getLegalIdentities().get(0), 99);
        CordaFuture<SignedTransaction> future = startFlow(a.getServices(), flow);
        network.runNetwork();
        SignedTransaction signedTransaction = future.get();

        assertEquals(1, signedTransaction.getTx().getOutputStates().size());
        TransactionState output = signedTransaction.getTx().getOutputs().get(0);

        assertEquals(network.getNotaryNodes().get(0).getInfo().getLegalIdentities().get(0), output.getNotary());
    }

    @Test
    public void transactionConstructedByFlowHasOneTokenStateOutputWithTheCorrectAmountAndRecipient() throws Exception {
        TokenFlow flow = new TokenFlow(b.getInfo().getLegalIdentities().get(0), 99);
        CordaFuture<SignedTransaction> future = startFlow(a.getServices(), flow);
        network.runNetwork();
        SignedTransaction signedTransaction = future.get();

        assertEquals(1, signedTransaction.getTx().getOutputStates().size());
        TokenState output = signedTransaction.getTx().outputsOfType(TokenState.class).get(0);

        assertEquals(b.getInfo().getLegalIdentities().get(0), output.getRecipient());
        assertEquals(99, output.getAmount());
    }

    @Test
    public void transactionConstructedByFlowHasOneOutputUsingTheCorrectContract() throws Exception {
        TokenFlow flow = new TokenFlow(b.getInfo().getLegalIdentities().get(0), 99);
        CordaFuture<SignedTransaction> future = startFlow(a.getServices(), flow);
        network.runNetwork();
        SignedTransaction signedTransaction = future.get();

        assertEquals(1, signedTransaction.getTx().getOutputStates().size());
        TransactionState output = signedTransaction.getTx().getOutputs().get(0);

        assertEquals("bootcamp.TokenContract", output.getContract());
    }

    @Test
    public void transactionConstructedByFlowHasOneIssueCommand() throws Exception {
        TokenFlow flow = new TokenFlow(b.getInfo().getLegalIdentities().get(0), 99);
        CordaFuture<SignedTransaction> future = startFlow(a.getServices(), flow);
        network.runNetwork();
        SignedTransaction signedTransaction = future.get();

        assertEquals(1, signedTransaction.getTx().getCommands().size());
        Command command = signedTransaction.getTx().getCommands().get(0);

        assert(command.getValue() instanceof TokenContract.Issue);
    }

    @Test
    public void transactionConstructedByFlowHasOneCommandWithTheIssueAsASigner() throws Exception {
        TokenFlow flow = new TokenFlow(b.getInfo().getLegalIdentities().get(0), 99);
        CordaFuture<SignedTransaction> future = startFlow(a.getServices(), flow);
        network.runNetwork();
        SignedTransaction signedTransaction = future.get();

        assertEquals(1, signedTransaction.getTx().getCommands().size());
        Command command = signedTransaction.getTx().getCommands().get(0);

        assertEquals(1, command.getSigners().size());
        assert(command.getSigners().contains(a.getInfo().getLegalIdentities().get(0).getOwningKey()));
    }

    @Test
    public void transactionConstructedByFlowHasNoInputsAttachmentsOrTimewindows() throws Exception {
        TokenFlow flow = new TokenFlow(b.getInfo().getLegalIdentities().get(0), 99);
        CordaFuture<SignedTransaction> future = startFlow(a.getServices(), flow);
        network.runNetwork();
        SignedTransaction signedTransaction = future.get();

        assertEquals(0, signedTransaction.getTx().getInputs().size());
        // The single attachment is the contract attachment.
        assertEquals(1, signedTransaction.getTx().getAttachments().size());
        assertEquals(null, signedTransaction.getTx().getTimeWindow());
    }
}

File: kotlin/src/test/kotlin/kotlin_bootcamp/FlowTests.kt
package kotlin_bootcamp

import net.corda.testing.internal.chooseIdentity
import net.corda.testing.node.MockNetwork
import net.corda.testing.node.StartedMockNode
import org.junit.After
import org.junit.Before
import org.junit.Test
import kotlin.test.assertEquals

class FlowTests {
    lateinit var network: MockNetwork
    lateinit var a: StartedMockNode
    lateinit var b: StartedMockNode

    @Before
    fun setup() {
        network = MockNetwork(listOf("kotlin_bootcamp"))
        a = network.createPartyNode()
        b = network.createPartyNode()
        network.runNetwork()
    }

    @After
    fun tearDown() {
        network.stopNodes()
    }

    @Test
    fun `transaction constructed by flow uses the correct notary`() {
        val flow = TokenFlow(b.info.chooseIdentity(), 99)
        val future = a.startFlow(flow)
        network.runNetwork()
        val signedTransaction = future.get()

        assertEquals(1, signedTransaction.tx.outputStates.size)
        val output = signedTransaction.tx.outputs.single()

        assertEquals(network.notaryNodes.single().info.chooseIdentity(), output.notary)
    }

    @Test
    fun `transaction constructed by flow has one TokenState output with the correct amount and recipient`() {
        val flow = TokenFlow(b.info.chooseIdentity(), 99)
        val future = a.startFlow(flow)
        network.runNetwork()
        val signedTransaction = future.get()

        assertEquals(1, signedTransaction.tx.outputStates.size)
        val output = signedTransaction.tx.outputsOfType<TokenState>().single()

        assertEquals(b.info.chooseIdentity(), output.recipient)
        assertEquals(99, output.amount)
    }

    @Test
    fun `transaction constructed by flow has one output using the correct contract`() {
        val flow = TokenFlow(b.info.chooseIdentity(), 99)
        val future = a.startFlow(flow)
        network.runNetwork()
        val signedTransaction = future.get()

        assertEquals(1, signedTransaction.tx.outputStates.size)
        val output = signedTransaction.tx.outputs.single()

        assertEquals("kotlin_bootcamp.TokenContract", output.contract)
    }

    @Test
    fun `transaction constructed by flow has one Issue command`() {
        val flow = TokenFlow(b.info.chooseIdentity(), 99)
        val future = a.startFlow(flow)
        network.runNetwork()
        val signedTransaction = future.get()

        assertEquals(1, signedTransaction.tx.commands.size)
        val command = signedTransaction.tx.commands.single()

        assert(command.value is TokenContract.Issue)
    }

    @Test
    fun `transaction constructed by flow has one command with the issuer as a signer`() {
        val flow = TokenFlow(b.info.chooseIdentity(), 99)
        val future = a.startFlow(flow)
        network.runNetwork()
        val signedTransaction = future.get()

        assertEquals(1, signedTransaction.tx.commands.size)
        val command = signedTransaction.tx.commands.single()

        assertEquals(setOf(a.info.chooseIdentity().owningKey), command.signers.toSet())
    }

    @Test
    fun `transaction constructed by flow has no inputs, attachments or timewindows`() {
        val flow = TokenFlow(b.info.chooseIdentity(), 99)
        val future = a.startFlow(flow)
        network.runNetwork()
        val signedTransaction = future.get()

        assertEquals(0, signedTransaction.tx.inputs.size)
        // The single attachment is the contract attachment.
        assertEquals(1, signedTransaction.tx.attachments.size)
        assertEquals(null, signedTransaction.tx.timeWindow)
    }
}